### ISSUE TITLE
Building without httpd is broken

### ISSUE THREAD

**Issue by dilyanpalauzov** [cyrusimap/cyrus-imapd#2754]
Since 4464040cebea114dfe imap/http_client.c is not anymore part only of imap/httpd, but is included in libcyrus_imap.la . imap/http_client.c \#imports "imap/http_err.h", which is only built, when --enable-httpd was passed to ./configure .

When ./configure was called without --enable-httpd, compliling libimap fails.

**Comment by dilyanpalauzov** [comment]
imap/conversations.c is unconditinal part of libcyrus_imap.la .  conversations.c uses unconditionally record_to_vcard() and vcard_prop_decode_value().  The latter are defined in imap/vcard_support.c, but vcard_support.c is added to libcyrus_imap, only when the http server is build.

When no http server is build, libcyrus_imap contains undefined symbols.

**Comment by ksmurchison** [comment]
This seems to fix it for me:

https://github.com/cyrusimap/cyrus-imapd/commit/c70e3dacdbf7405bf8fd895ba39a98eaa6381c54

I don't have any issue with http_err.h when not building with httpd

**Comment by dilyanpalauzov** [comment]
The file http_err.c is only needed, when httpd is built.  At https://github.com/cyrusimap/cyrus-imapd/pull/2711/ I have adjusted Makefile.am to create http_err.c and http_err.h only when httpd is build.

Is there a use-case, where http_err.h is needed for compiling, but http_err.c is not?

**Comment by elliefm** [comment]
#2711 is merged now, but the build still fails when configured with `--disable-http`

I'm digging through commit history and it looks like things went like this:
* we added a curl dependency for some functionality
* we moved http_client.c from httpd to libcyrus_imap.la so that we could use that instead of curl, and ripped out the curl dependency

http_client.c just flat out depends on http_err because it needs to be able to return HTTP_ error codes, so the only coherent fix here is to move http_err into libcyrus_imap.so too.  Testing a patch for that at the moment...

**Comment by elliefm** [comment]
Okay, this is not quite as trivial as it seemed.  

The HTTP_ error codes work fine when http_err.c is in libcyrus_imap.la rather than httpd, because those are just defines.

But http_proxy wants to examine `et_http_error_table`.  This is fine when http_err is built right into httpd, but when it's in libcyrusimap.la it's invisible, because it hasn't been marked with the EXPORTED tag:

```
$nm imap/.libs/libcyrus_imap.so | grep 'et_.*_error_table'
0000000000354990 d et_http_error_table
0000000000354c00 d et_imap_error_table
0000000000354c70 d et_mupd_error_table
```

I believe we need these to be "D" (not "d") for http_proxy.c to see it the way it wants to when it's bundled in the library.  I'm not sure if there's a way to persuade compile_et to generate the file with the correct visibility, I haven't looked yet.

My work-in-progress is here: https://github.com/elliefm/cyrus-imapd/tree/v31/2754-move-http_err-into-libcyrus_imap

It seems to fix the --disable-http build, but breaks the --enable-http build

**Comment by elliefm** [comment]
> Is there a use-case, where http_err.h is needed for compiling, but http_err.c is not?

Hmm... if http_client.c _never_ calls error_message() for one of the HTTP_ errors, then I think maybe it can get by without http_err.c and just use the defines from http_err.h

In which case, we would still need to always build http_err.[ch] to ensure we have the header even if we don't have httpd, but we could get away with leaving http_err.c out of libcyrus_imap.la and only building it into httpd.  Which would fix the visibility problem my current patch introduces in http_proxy.c.

But that feels so precarious, how long until we trip over it again and end up having to fix it properly anyway??

**Comment by dilyanpalauzov** [comment]
Ideally, the codepart of http_client.c that uses http_err.h shall be changed, so that it is compiled only when the error codes will actually be used - when HTTPD is build.  This could mean splitting http_client.c in two pieces: one general, and one condititonal.

You can try setting “extern EXPORTED et_http_error_table” in any file from libcyrus_imap and see if the linker applies the visibility for the whole library.

**Comment by elliefm** [comment]
> You can try setting “extern EXPORTED et_http_error_table” in any file from libcyrus_imap and see if the linker applies the visibility for the whole library.

Hmm. I've just tried something like this from global.c, but it wasn't enough.

Also tried (again, in global.c):

```
#pragma GCC visibility push(default)
extern const struct error_table et_http_error_table;
#pragma GCC visibility pop
```

but that didn't help either.

Manually editing http_err.c to inject the attribute:

```
const struct error_table et_http_error_table __attribute__((__visibility__("default"))) = { text, -1964267008L, 64 };
```

does work, though.  And nm confirms it (note 'D'):

```
$nm imap/.libs/libcyrus_imap.so | grep 'et_.*_error_table'
0000000000354990 D et_http_error_table
0000000000354c00 d et_imap_error_table
0000000000354c70 d et_mupd_error_table
```

So the linkage needed here _can_ be achieved by a sed command if nothing better turns up...

**Comment by elliefm** [comment]
Turns out compile_et is actually just a bunch of awk scripts that do a lot of string mangling.  They have no particular understanding of C syntax and therefore don't (and can't) have options for symbol linkage.

My branch above now has a commit to Makefile.am that mangles them further after they're generated (i.e. the same edit I did by hand in my previous comment).  It works for me, but it might break on builds that use a system com_err instead of the one we ship, or that provide an inadequate sed utility.  I don't like it, it feels precarious.

So, now I'm looking more closely at the http_proxy.c code that needs access to this table...  It's only needed by `http_status_to_code(unsigned code)`, which, if I'm reading it right, takes a numeric HTTP status (e.g. `404`) and converts it into the corresponding com_err error code (e.g. `-1964266981L`).  Seems like we could just move this function into libcyrus_imap.so as well (and change it from HIDDEN to EXPORTED), and our visibility problems would go away.

**Comment by elliefm** [comment]
In fact, that's exactly what I'm going to do.  Checking further, every time `http_status_to_code()` is called, it's directly after a call to `http_read_response()` (which provides the HTTP status to be converted), so it seems like a fundamental part of the http_client API, not a hidden function in the proxy :)

**Comment by elliefm** [comment]
Okay I think this is now properly fixed on master by this series of commits: 13938946b...e233cdf08 (which are the same as my branch but with the experimentation cleaned up)

Builds and tests cleanly both with and without the HTTP features built in.

(There's also a new commit on the Cassandane repo because the Prometheus tests didn't flag themselves as dependent on httpd, but they now do.)

**Comment by dilyanpalauzov** [comment]
As building _err.c implies building _err.c, there is no need to include _err.h in BUILD_SOURCES.

Adding .h to _SOURCES serves only including the file on "make dist".

Please remove the _err.h files from BUILT_SOURCES and (nodist)*_SOURCES.
